### ISSUE TITLE
Alias handling for "magic" projects broken

### ISSUE THREAD

**Issue by marvin2k** [rock-core/autoproj#88]
And for today: Did you know that calling
```bash
autoproj build typelib
```
is silently different from calling
```bash
autoproj build tools/typelib
```
Why the fuck?

Look at the diff of the two `CMakeCache.txt`...

Super idea having magic handling for "special" stuff in a general tool...

**Comment by doudou** [comment]
> Why the fuck?

Ever heard of "bugs" ? And maybe losing the attitude, that would be great.

> Super idea having magic handling for "special" stuff in a general tool...

Begging your pardon ? Grep for typelib in the autoproj source code and check whether there is anything special about it.

Can you actually make an actual bug report ? I don't even know what you mean by "silently different" here.

**Comment by marvin2k** [comment]
Hmpf. Sorry if I went to far... It's hard to stay focused in the task at hand when other things (like this issue) crop up along the way... Took some time to realize that specific behavior. And the actual bug-report is more or less complete. The diff for the typelib CMakeCache.txt is as follows (and can be easily reproduced):

```diff
--- build_back/CMakeCache.txt   2015-08-21 13:57:37.320443708 +0200
+++ build/CMakeCache.txt        2015-08-21 13:57:51.360513329 +0200
@@ -253,9 +253,6 @@
 //Path to a program.
 GITCOMMAND:FILEPATH=/usr/bin/git
 
-//No help, variable specified on the command line.
-GLIBC_HAVE_LONG_LONG:UNINITIALIZED=ON
-
 //Path to a program.
 HGCOMMAND:FILEPATH=HGCOMMAND-NOTFOUND
 
@@ -293,7 +290,7 @@
 RUBY_CONFIG_INCLUDE_DIR:PATH=/usr/include/x86_64-linux-gnu/ruby-2.1.0
 
 //Path to a program.
-RUBY_EXECUTABLE:FILEPATH=/usr/bin/ruby2.1
+RUBY_EXECUTABLE:FILEPATH=/home/mzenzes/rock_gccxml_master/install/bin/ruby
 
 //Vendor Ruby is available
 RUBY_HAS_VENDOR_RUBY:BOOL=
@@ -561,7 +558,7 @@
 //Details about finding PkgConfig
 FIND_PACKAGE_MESSAGE_DETAILS_PkgConfig:INTERNAL=[/usr/bin/pkg-config][v0.28()]
 //Details about finding Ruby
-FIND_PACKAGE_MESSAGE_DETAILS_Ruby:INTERNAL=[/usr/bin/ruby2.1][/usr/include/ruby-2.1.0][/usr/lib/x86_64-linux-gnu/libruby-2.1.so][/usr/include/x86_64-linux-gnu/ruby-2.1.0][v()]
+FIND_PACKAGE_MESSAGE_DETAILS_Ruby:INTERNAL=[/home/mzenzes/rock_gccxml_master/install/bin/ruby][/usr/include/ruby-2.1.0][/usr/lib/x86_64-linux-gnu/libruby-2.1.so][/usr/include/x86_64-linux-gnu/ruby-2.1.0][v()]
 //ADVANCED property for variable: GITCOMMAND
 GITCOMMAND-ADVANCED:INTERNAL=1
 //ADVANCED property for variable: HGCOMMAND
```
> Grep for typelib in the autoproj source code and check whether there is anything special about it.

As autoproj actually executes the code in "configuration" files, the border of the tool is not so sharp. Here, on my desk, the executed process triggered by calling `autoproj` makes this special handling. If this project here is the wrong place to note the issue: How should I know?

**Comment by goldhoorn** [comment]
I assume this is related to the problem when using a 'wrong' name in overrides.yml

this does not work:
```
- tools/orogen
  branch: something
```
a user must use:
```
- orogen
  branch: something
```

**Comment by doudou** [comment]
May I know what is wrong with that diff ? What I see is the standard cmake configuration for all packages built within Rock (RUBY_EXECUTABLE points to the autoproj ruby shim to avoid version conflicts for people with multiple ruby installed, GLIBC_HAVE_LONG_LONG is a workaround that had been added by steffen planthaber for some building issues on debian).

Here's the bottom line: if you want to hand-tune every cmake configuration in your install *don't use autoproj ever*. But, also, good luck with trying to reproduce builds.

**Comment by doudou** [comment]
> It's hard to stay focused in the task at hand when other things (like this issue) crop up along the way

Welcome to my world. Every time I start using a core part of Rock I spend 1/3 of my time fixing bugs. <non-constructive rant>But, of course, it is so much more important to spend a lot of developer resources in developing a new toolchain in C++. Polishing is for dumb people</non-constructive rant>

**Comment by goldhoorn** [comment]
Let me rephrase the problem:
Martin tried to build typelib once by calling typelib and once by calling tools/typelib.

The most relevant part is that autoproj sets NO define at all, the core problem is:
``` 18 BUILD_CLANG_TLB_IMPORTER:BOOL=OFF ``` which martin has left out.

**Comment by marvin2k** [comment]
> Polishing is for dumb people

I think we agree here. Polishing is very important, and too much unpolished edges drives users and developers away.

Well, as @goldhoorn said: The problem is when trying to enable `clang` by doing
```diff
diff --git a/init.rb b/init.rb
index 9f642e9..bbb7ad1 100644
--- a/init.rb
+++ b/init.rb
@@ -24,3 +24,6 @@
 
 require 'autoproj/gitorious'
 Autoproj.gitorious_server_configuration('DFKIGIT', 'git.hb.dfki.de', :http_url => 'https://git.hb.dfki.de')
+
+Autoproj.env_set 'TYPELIB_CXX_LOADER', 'clang'
```
which will somehow not work when calling `autoproj build tools/typelib`. This might be fixed by doubling the relevant block in [orocos.autobuild](https://github.com/orocos-toolchain/autoproj/blob/master/orocos.autobuild#L87)? But the underlying problem prevails: This "aliasing" is not working properly?

**Comment by doudou** [comment]
> Autoproj.env_set 'TYPELIB_CXX_LOADER', 'clang'

This sets an environment variable which is meant to be propagated to the cmake configuration (TYPELIB_CXX_LOADER). Your diff says nothing about TYPELIB_CXX_LOADER...

**Comment by marvin2k** [comment]
Yes. I left this part out to make the diff smaller. It does not matter if the `clang` is part of the diff -- the two calls result in different configurations for cmake.

**Comment by doudou** [comment]
OK ... did not understand all of it. Which is the one with 'typelib' and which is with 'tools/typelib'

**Comment by marvin2k** [comment]
Right, should've named the two CMakeCaches more distinctly... And should not have left the clang-part. So here a new diff:
```diff
--- autoproj-build-tools-typelibs/CMakeCache.txt	2015-08-21 15:12:45.834800224 +0200
+++ autoproj-build-typelib/CMakeCache.txt	2015-08-21 15:11:56.654556352 +0200
@@ -15,7 +15,7 @@
 ########################
 
 //set to ON to enable compiling of the llvm/clang based importer
-BUILD_CLANG_TLB_IMPORTER:BOOL=OFF
+BUILD_CLANG_TLB_IMPORTER:BOOL=ON
 
 //Build the testing tree.
 BUILD_TESTING:BOOL=ON
@@ -253,12 +253,24 @@
 //Path to a program.
 GITCOMMAND:FILEPATH=/usr/bin/git
 
+//No help, variable specified on the command line.
+GLIBC_HAVE_LONG_LONG:UNINITIALIZED=ON
+
 //Path to a program.
 HGCOMMAND:FILEPATH=HGCOMMAND-NOTFOUND
 
+//Path to a file.
+LIBDL_INCLUDE_DIR:PATH=/usr/include
+
+//Path to a library.
+LIBDL_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libdl.so
+
 //Single output directory for building all libraries.
 LIBRARY_OUTPUT_PATH:PATH=
 
+//Path to a program.
+LLVM_CONFIG_EXECUTABLE:FILEPATH=/usr/bin/llvm-config-3.4
+
 //Command to build the project
 MAKECOMMAND:STRING=/usr/bin/cmake --build . --config "${CTEST_CONFIGURATION_TYPE}" -- -i
 
@@ -290,7 +302,7 @@
 RUBY_CONFIG_INCLUDE_DIR:PATH=/usr/include/x86_64-linux-gnu/ruby-2.1.0
 
 //Path to a program.
-RUBY_EXECUTABLE:FILEPATH=/home/mzenzes/rock_clang_master/install/bin/ruby
+RUBY_EXECUTABLE:FILEPATH=/usr/bin/ruby2.1
 
 //Vendor Ruby is available
 RUBY_HAS_VENDOR_RUBY:BOOL=
@@ -350,6 +362,9 @@
 //Path to a program.
 YARD_EXECUTABLE:FILEPATH=/home/mzenzes/rock_clang_master/.gems/bin/yard
 
+//Path to a library.
+llvm_lib_path:FILEPATH=/usr/lib/llvm-3.4/lib/libLLVMSupport.a
+
 //Dependencies for the target
 typeLang_cSupport_LIB_DEPENDS:STATIC=general;typeLib;
 
@@ -472,6 +487,14 @@
 CMAKE_GENERATOR:INTERNAL=Unix Makefiles
 //Name of generator toolset.
 CMAKE_GENERATOR_TOOLSET:INTERNAL=
+//Have symbol pthread_create
+CMAKE_HAVE_LIBC_CREATE:INTERNAL=
+//Have library pthreads
+CMAKE_HAVE_PTHREADS_CREATE:INTERNAL=
+//Have library pthread
+CMAKE_HAVE_PTHREAD_CREATE:INTERNAL=1
+//Have include pthread.h
+CMAKE_HAVE_PTHREAD_H:INTERNAL=1
 //Start directory with the top level CMakeLists.txt file for this
 // project
 CMAKE_HOME_DIRECTORY:INTERNAL=/home/mzenzes/rock_clang_master/tools/typelib
@@ -494,7 +517,7 @@
 //ADVANCED property for variable: CMAKE_NM
 CMAKE_NM-ADVANCED:INTERNAL=1
 //number of local generators
-CMAKE_NUMBER_OF_LOCAL_GENERATORS:INTERNAL=11
+CMAKE_NUMBER_OF_LOCAL_GENERATORS:INTERNAL=13
 //ADVANCED property for variable: CMAKE_OBJCOPY
 CMAKE_OBJCOPY-ADVANCED:INTERNAL=1
 //ADVANCED property for variable: CMAKE_OBJDUMP
@@ -555,14 +578,22 @@
 DOXYGEN_DOT_PATH-ADVANCED:INTERNAL=1
 //ADVANCED property for variable: DOXYGEN_EXECUTABLE
 DOXYGEN_EXECUTABLE-ADVANCED:INTERNAL=1
+//Details about finding LibDL
+FIND_PACKAGE_MESSAGE_DETAILS_LibDL:INTERNAL=[/usr/lib/x86_64-linux-gnu/libdl.so][/usr/include][v()]
 //Details about finding PkgConfig
 FIND_PACKAGE_MESSAGE_DETAILS_PkgConfig:INTERNAL=[/usr/bin/pkg-config][v0.28()]
 //Details about finding Ruby
-FIND_PACKAGE_MESSAGE_DETAILS_Ruby:INTERNAL=[/home/mzenzes/rock_clang_master/install/bin/ruby][/usr/include/ruby-2.1.0][/usr/lib/x86_64-linux-gnu/libruby-2.1.so][/usr/include/x86_64-linux-gnu/ruby-2.1.0][v()]
+FIND_PACKAGE_MESSAGE_DETAILS_Ruby:INTERNAL=[/usr/bin/ruby2.1][/usr/include/ruby-2.1.0][/usr/lib/x86_64-linux-gnu/libruby-2.1.so][/usr/include/x86_64-linux-gnu/ruby-2.1.0][v()]
+//Details about finding Threads
+FIND_PACKAGE_MESSAGE_DETAILS_Threads:INTERNAL=[TRUE][v()]
 //ADVANCED property for variable: GITCOMMAND
 GITCOMMAND-ADVANCED:INTERNAL=1
 //ADVANCED property for variable: HGCOMMAND
 HGCOMMAND-ADVANCED:INTERNAL=1
+//ADVANCED property for variable: LIBDL_INCLUDE_DIR
+LIBDL_INCLUDE_DIR-ADVANCED:INTERNAL=1
+//ADVANCED property for variable: LIBDL_LIBRARY
+LIBDL_LIBRARY-ADVANCED:INTERNAL=1
 LibXML_CFLAGS:INTERNAL=-I/usr/include/libxml2
 LibXML_CFLAGS_I:INTERNAL=
 LibXML_CFLAGS_OTHER:INTERNAL=
@@ -649,5 +680,40 @@
 SLURM_SRUN_COMMAND-ADVANCED:INTERNAL=1
 //ADVANCED property for variable: SVNCOMMAND
 SVNCOMMAND-ADVANCED:INTERNAL=1
+TINFO_CFLAGS:INTERNAL=-I/usr/include/ncursesw
+TINFO_CFLAGS_I:INTERNAL=
+TINFO_CFLAGS_OTHER:INTERNAL=
+TINFO_FOUND:INTERNAL=1
+TINFO_INCLUDEDIR:INTERNAL=/usr/include/ncursesw
+TINFO_INCLUDE_DIRS:INTERNAL=/usr/include/ncursesw
+TINFO_LDFLAGS:INTERNAL=-ltinfo
+TINFO_LDFLAGS_OTHER:INTERNAL=
+TINFO_LIBDIR:INTERNAL=/usr/lib/x86_64-linux-gnu
+TINFO_LIBRARIES:INTERNAL=tinfo
+TINFO_LIBRARY_DIRS:INTERNAL=
+TINFO_LIBS:INTERNAL=
+TINFO_LIBS_L:INTERNAL=
+TINFO_LIBS_OTHER:INTERNAL=
+TINFO_LIBS_PATHS:INTERNAL=
+TINFO_PREFIX:INTERNAL=/usr
+TINFO_STATIC_CFLAGS:INTERNAL=-I/usr/include/ncursesw
+TINFO_STATIC_CFLAGS_I:INTERNAL=
+TINFO_STATIC_CFLAGS_OTHER:INTERNAL=
+TINFO_STATIC_INCLUDE_DIRS:INTERNAL=/usr/include/ncursesw
+TINFO_STATIC_LDFLAGS:INTERNAL=-ltinfo
+TINFO_STATIC_LDFLAGS_OTHER:INTERNAL=
+TINFO_STATIC_LIBDIR:INTERNAL=
+TINFO_STATIC_LIBRARIES:INTERNAL=tinfo
+TINFO_STATIC_LIBRARY_DIRS:INTERNAL=
+TINFO_STATIC_LIBS:INTERNAL=
+TINFO_STATIC_LIBS_L:INTERNAL=
+TINFO_STATIC_LIBS_OTHER:INTERNAL=
+TINFO_STATIC_LIBS_PATHS:INTERNAL=
+TINFO_VERSION:INTERNAL=5.9.20140913
+TINFO_tinfo_INCLUDEDIR:INTERNAL=
+TINFO_tinfo_LIBDIR:INTERNAL=
+TINFO_tinfo_PREFIX:INTERNAL=
+TINFO_tinfo_VERSION:INTERNAL=
 __pkg_config_checked_LibXML:INTERNAL=1
+__pkg_config_checked_TINFO:INTERNAL=1
```

**Comment by doudou** [comment]
And this happens every time you switch between the two command lines ?

**Comment by marvin2k** [comment]
Yes, as far as I can see.

**Comment by doudou** [comment]
OK ... I have to bootstrap a clean install that is not using autoproj 2.0 beta ...

**Comment by doudou** [comment]
Forgot that one ... Anyways.

I did a clean test today with 1.13.3 and 2.0 RC and both do not show this behaviour. Looking at it again, it smells like autoproj would auto-discover tools/typelib instead of matching it against the typelib package when you ran `autoproj build tools/typelib`. If it ever happens again, check the autoproj output for a line looking like `auto-adding tools/typelib using the cmake handler` (or something approaching) and definitely re-open this issue.

**Comment by marvin2k** [comment]
Hm, I observed something similar today: Having the following lines added to "overrides.yml":

```yaml
    - tools/orogen:
         branch: master
```

and then calling `aup tools/orogen` does still use the default-branch, nothing changes. But using the override

```yaml
    - orogen:
         branch: master
```

and then calling `aup tools/orogen` works as expected: `master` is checked out. This behavior does not occur with the `typelib` package. Don't know if this is related to the OP, but when I saw this today, this issue came back to my mind ;-)

There is no such lines as "auto-adding something" during the whole lot of output. Autoproj is v1.13.4, system is rock-stable on Debian Jessie with handpicked parts from master. Ruby is 2.1.5

**Comment by doudou** [comment]
The package name and the place it is in the tree do not necessarily match in autoproj (you can place a package in an arbitrary place in the source tree). The real issue here is that there is no validation of package names in overrides.

**Comment by marvin2k** [comment]
Hm. Well. Will now hijack the thread ;-) Maybe you can also guess what going wrong here:

```yaml
  - rtt:
      github:  john_doe/rtt
      branch: nice_shiny_feature
      tag: ~
```

remote of rtt is changed, but the branch is not. Feeling dumb.

**Comment by doudou** [comment]
What does 'autoproj show rtt' say ?

**Comment by marvin2k** [comment]
Ah, yes. There the output is correct (branchname _and_ remote)

```
source package rtt
  source definition
    type: git
    url: https://github.com/john_doe/rtt.git
    branch: nice-shiny-feature
    push_to: git@github.com:/john_doe/rtt.git
    repository_id: github:/john_doe/rtt.git
    retry_count: 10
    tag: 
    first match: in orocos.toolchain
      github: orocos-toolchain/$PACKAGE
    overriden in main configuration
      branch: nice-shiny-feature
      github: john_doe/rtt
      tag: 
<cropped for sanity>
```
